File: backend/src/services/dashboard.service.ts
import { prisma } from "../config";

export async function getDashboard(userId: string) {
  const [
    totalTasks,
    completedTasks,
    openIssues,
    totalFeedback,
    totalNotes,
    recentTasks,
    recentIssues,
    recentFeedback,
    recentNotes,
  ] = await Promise.all([
    prisma.task.count({ where: { userId } }),
    prisma.task.count({ where: { userId, status: "completed" } }),
    prisma.issue.count({ where: { userId, status: { in: ["open", "in_progress"] } } }),
    prisma.feedback.count({ where: { userId } }),
    prisma.note.count({ where: { userId } }),
    prisma.task.findMany({
      where: { userId },
      select: { id: true, date: true, title: true, status: true, priority: true, category: true },
      orderBy: { date: "desc" },
      take: 5,
    }),
    prisma.issue.findMany({
      where: { userId },
      select: { id: true, date: true, title: true, status: true, severity: true },
      orderBy: { date: "desc" },
      take: 5,
    }),
    prisma.feedback.findMany({
      where: { userId },
      select: { id: true, date: true, subject: true, type: true },
      orderBy: { date: "desc" },
      take: 5,
    }),
    prisma.note.findMany({
      where: { userId },
      select: { id: true, date: true, title: true, tags: true },
      orderBy: { date: "desc" },
      take: 5,
    }),
  ]);

  const taskCompletionRate = totalTasks > 0 ? Math.round((completedTasks / totalTasks) * 100) : 0;

  return {
    summary: {
      totalTasks,
      completedTasks,
      openIssues,
      totalFeedback,
      totalNotes,
      taskCompletionRate,
    },
    recentTasks,
    recentIssues,
    recentFeedback,
    recentNotes,
  };
}

File: dot-net-backend/src/OnboardingDiary.Api/Services/DashboardService.cs
using Microsoft.EntityFrameworkCore;
using OnboardingDiary.Api.Data;
using OnboardingDiary.Api.DTOs.Dashboard;
using OnboardingDiary.Api.Models.Enums;

namespace OnboardingDiary.Api.Services;

public class DashboardService : IDashboardService
{
    private readonly AppDbContext _db;

    public DashboardService(AppDbContext db) => _db = db;

    public async Task<DashboardDto> GetDashboardAsync(Guid userId)
    {
        var tasks = _db.TaskEntries.Where(t => t.UserId == userId);
        var issues = _db.IssueEntries.Where(i => i.UserId == userId);

        var totalTasks = await tasks.CountAsync();
        var completedTasks = await tasks.CountAsync(t => t.Status == Models.Enums.TaskStatus.Completed);
        var inProgressTasks = await tasks.CountAsync(t => t.Status == Models.Enums.TaskStatus.InProgress);

        var totalIssues = await issues.CountAsync();
        var openIssues = await issues.CountAsync(i => i.Status == IssueStatus.Open);
        var resolvedIssues = await issues.CountAsync(i => i.Status == IssueStatus.Resolved);

        var totalFeedback = await _db.FeedbackEntries.CountAsync(f => f.UserId == userId);
        var totalNotes = await _db.NoteEntries.CountAsync(n => n.UserId == userId);

        var recentTasks = await tasks.OrderByDescending(t => t.CreatedAt).Take(5)
            .Select(t => new RecentEntryDto { Id = t.Id, Type = "Task", Title = t.Title, Date = t.Date, CreatedAt = t.CreatedAt })
            .ToListAsync();

        var recentIssues = await issues.OrderByDescending(i => i.CreatedAt).Take(5)
            .Select(i => new RecentEntryDto { Id = i.Id, Type = "Issue", Title = i.Title, Date = i.Date, CreatedAt = i.CreatedAt })
            .ToListAsync();

        var recentActivity = recentTasks.Concat(recentIssues)
            .OrderByDescending(r => r.CreatedAt).Take(10).ToList();

        return new DashboardDto
        {
            TotalTasks = totalTasks,
            CompletedTasks = completedTasks,
            InProgressTasks = inProgressTasks,
            TotalIssues = totalIssues,
            OpenIssues = openIssues,
            ResolvedIssues = resolvedIssues,
            TotalFeedback = totalFeedback,
            TotalNotes = totalNotes,
            TaskCompletionRate = totalTasks > 0 ? Math.Round((double)completedTasks / totalTasks * 100, 1) : 0,
            RecentActivity = recentActivity,
        };
    }
}
